kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/Utilities.java

```diff
@@ -0,0 +1,254 @@
+    }
+  }
+
+  private static Object getValueFromStruct(Struct struct, String[] fields, int idx) {
+    Preconditions.checkArgument(idx < fields.length, "Invalid field index");
+    Object value = struct.get(fields[idx]);
+    if (value == null || idx == fields.length - 1) {
+      return value;
+    }
+
+    Preconditions.checkState(value instanceof Struct, "Expected a struct type");
+    return getValueFromStruct((Struct) value, fields, idx + 1);
+  }
+
+  private static Object getValueFromMap(Map<?, ?> map, String[] fields, int idx) {
+    Preconditions.checkArgument(idx < fields.length, "Invalid field index");
+    Object value = map.get(fields[idx]);
+    if (value == null || idx == fields.length - 1) {
+      return value;
+    }
+
+    Preconditions.checkState(value instanceof Map, "Expected a map type");
+    return getValueFromMap((Map<?, ?>) value, fields, idx + 1);
+  }
+
+  public static TaskWriter<Record> createTableWriter(
+      Table table, String tableName, IcebergSinkConfig config) {
+    Map<String, String> tableProps = Maps.newHashMap(table.properties());
+    tableProps.putAll(config.writeProps());
+
+    String formatStr = tableProps.getOrDefault(DEFAULT_FILE_FORMAT, DEFAULT_FILE_FORMAT_DEFAULT);
+    FileFormat format = FileFormat.valueOf(formatStr.toUpperCase());
+
+    long targetFileSize =
+        PropertyUtil.propertyAsLong(
+            tableProps, WRITE_TARGET_FILE_SIZE_BYTES, WRITE_TARGET_FILE_SIZE_BYTES_DEFAULT);
+
+    Set<Integer> identifierFieldIds = table.schema().identifierFieldIds();
+
+    // override the identifier fields if the config is set
+    List<String> idCols = config.tableConfig(tableName).idColumns();
+    if (!idCols.isEmpty()) {
+      identifierFieldIds =
+          idCols.stream()
+              .map(colName -> table.schema().findField(colName).fieldId())
+              .collect(toSet());
+    }
+
+    FileAppenderFactory<Record> appenderFactory;
+    if (identifierFieldIds == null || identifierFieldIds.isEmpty()) {
+      appenderFactory =
+          new GenericAppenderFactory(table.schema(), table.spec(), null, null, null)
+              .setAll(tableProps);
+    } else {
+      appenderFactory =
+          new GenericAppenderFactory(
+                  table.schema(),
+                  table.spec(),
+                  Ints.toArray(identifierFieldIds),
+                  TypeUtil.select(table.schema(), Sets.newHashSet(identifierFieldIds)),
+                  null)
+              .setAll(tableProps);
+    }
+
+    // (partition ID + task ID + operation ID) must be unique
+    OutputFileFactory fileFactory =
+        OutputFileFactory.builderFor(table, 1, System.currentTimeMillis())
+            .defaultSpec(table.spec())
+            .operationId(UUID.randomUUID().toString())
+            .format(format)
+            .build();
+
+    TaskWriter<Record> writer;
+    if (table.spec().isUnpartitioned()) {
+      if (config.tablesCdcField() == null && !config.upsertModeEnabled()) {
```

Comment:
Is it practical to remove the row-level deletes for a future commit? It would be much easier to get this in if we focused on the append use case first.

Reply:
Yes I removed the delta writers for this PR.